src/libpmemobj/heap.c

```diff
@@ -944,10 +1030,64 @@ heap_get_narenas(struct palloc_heap *heap)
 struct bucket **
 heap_get_arena_buckets(struct palloc_heap *heap, unsigned arena_id)
 {
-	struct arena *a = &heap->rt->arenas[arena_id];
+	struct arena *a = VEC_ARR(&heap->rt->arenas)[arena_id];
 	return a->buckets;
 }
 
+/*
+ * heap_get_arena_auto -- returns arena automatic value
+ */
+int
+heap_get_arena_auto(struct palloc_heap *heap, unsigned arena_id)
+{
+	struct heap_rt *h = heap->rt;
+
+	util_mutex_lock(&h->arenas_lock);
+	struct arena *a = VEC_ARR(&heap->rt->arenas)[arena_id];
+	util_mutex_unlock(&h->arenas_lock);
+
+	return a->automatic;
+}
+
+/*
+ * heap_set_arena_auto -- sets arena automatic value
+ */
+void
+heap_set_arena_auto(struct palloc_heap *heap, unsigned arena_id,
+		int automatic)
+{
+	struct heap_rt *h = heap->rt;
+
+	util_mutex_lock(&h->arenas_lock);
+	struct arena *a = VEC_ARR(&heap->rt->arenas)[arena_id];
+	a->automatic = automatic;
+	util_mutex_unlock(&h->arenas_lock);
+}
+
+/*
+ * heap_set_arena_thread -- assign arena to the current thread
+ */
+int
+heap_set_arena_thread(struct palloc_heap *heap, unsigned arena_id)
+{
+	struct heap_rt *h = heap->rt;
+
+	util_mutex_lock(&h->arenas_lock);
+
+	if (arena_id < 0 || arena_id >= (unsigned)VEC_SIZE(&h->arenas))
+		return -1;
+
+	struct arena *a = VEC_ARR(&heap->rt->arenas)[arena_id];
+	util_mutex_unlock(&h->arenas_lock);
+
+	if (a != NULL && a->nthreads < 1) {
+		util_fetch_and_add64(&a->nthreads, 1);
```

Comment:
This is racy - if two threads will try to assign the same arena they may both succeed.
Please use compare & swap operation for that.

When arena doesn't exist or assignment fails please return !0.

Reply:
Assigning two threads the same arena in parallel should succeed. Fetch and add is correct here.

It's the `a->nthreads < 1` that's the bug here ;P Please remove it.